VTGate V3: Feature summary

sougou opened the pull request:
@AndyDiamondstein @michael-berlin @erzel
I've renamed the old vtgatev3 doc to VTGateV3Design, which is not very interesting to someone who wants to use V3. This new doc fulfills the purpose.

AndyDiamondstein commented:
It might be more clear to name the new document VTGateV3Features.md or VTGateV3Overview.md or something else that will distinguish it from other VTGateV3... docs down the line (implementation, reference, codelab, etc.).

sougou commented:
Makes sense. Renamed.

erzel commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,77 @@
+In the case of an order_detail table, it may only need an order_id foreign key. Since this foreign key means the same thing as the order_id in order, creating a cross-shard index for it will result in a duplication of the order_id->customer_id index. In such situations, V3 allows you to just reuse the existing index for the order_detail table also. This saves disk space and also reduces the overa ...
+
+### Consistency
+Once you add multiple indexes to tables, it's possible that the application could make inconstent requests. V3 makes sure that none of the specified constraints are broken. For example, if a table had both a basic sharding key and a hashed sharding key, it will enforce the rule that the hash of the basic sharding key matches that of the hashed sharding key.
```
typo: inconstent

erzel commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,77 @@
+Vitess's filtered replication currently requires that the hash value be physically present as a column in each table. To satisfy this need, you still need to create a column to store this hash value. However, V3 will take care of populating this on your behalf. *We will soon be removing this restriction once we change filtered replication to also perform the same hashing.*
+
+#### Auto-increment columns
+When a table gets sharded, you are no longer able to use MySQL's auto increment functionality. V3 allows you to designate a table in an unsharded database as the source of auto-increment ids. Once you've specified this, V3 will transparently use generated values from this table to keep the auto-increment going. Additionally, this column can be a hashed sharding key. This means that the insert wil ...
```
“Additionally, this column can be a hashed sharding key”. Did you mean the source of a hashed sharding key; i.e. the hash function is operated on this auto-incremented column, right? If so, can you clarify the language.

erzel commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,77 @@
+
+### Query diversity
+V3 does not support the full SQL feature set. The current implementation supports simple queries:
+* Single table DML statements: This is a vitess-wide restriction where you can affect only one table and one sharding key per statement. *This restriction may be limited in the future.*
```
“limited in the future”. You probably mean “lifted in the future”.

erzel commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,77 @@
+#### Cross-shard indexes
+As your application evolves, you'll invariably find yourself wanting to fetch rows based on columns other than the main sharding key. For example, if you've sharded your database by user id, you may still want to be able find users by their username. If you only had the sharding key, such queries can only be answered by sending it to all shards. This could become very expensive as the number of s ...
+
+The typical strategy to address this problem is to build a separate lookup table and keep it up-to-date. In the above case, you may build a separate username->user_id relationship table. Once you've informed V3 of this table, it will know what to do with a query like 'select * from user where username=:value'. You can also configure V3 to keep this table up-to-date as you insert or delete data. I ...
```
I like the example here, it makes it easier to understand.

michael-berlin replied:
I would prefer if you use the customer_id and order_id example here as well.

This way, the "Shared indexes" section can reuse the example and extend it.

erzel commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,77 @@
+SQL is a very powerful language. You can build queries that can result in large amount of work and memory consumption involving big intermediate results. Such constructs where the scope of work is open-ended will not be immediately supported. In such cases, it's recommended that you use map-reduce techniques for which there is a separate API. *We'll consider building such on-the-fly map-reducers  ...
+
+## The vschema editor
+The above features require metadata like sharding key, cross-shard indexes, etc. to be configured and stored in some place. This is known as the vschema. *We'll be building a vschema editor and wizard that will allow you to easily view and modify this metadata.*
```
I'd rephrase as: "metadata like configuration of sharding key and cross-shard indexes" to make it clear that it's the configuration information that is being stored in the vschema not the entities themselves.

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+# VTGate V3 features
+
+## Overview
```
As very first sentence, I would like to see here a definition of vtgate first. This will give a better introduction.

For me, vtgate is a proxy which contains the routing logic. You can also mention here that it's basically a fat client, but provided as a service. Over time, we moved away from a fat client, which had to be incorporated into the application, to the proxy model. This way, the routing logic is more decoupled from the application and easier to maintain separately.

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+## Feature set
+The [V3 design](https://github.com/youtube/vitess/blob/master/doc/VTGateV3.md) is quite elaborate. If necessary, it will allow you to plug in custom indexes and sharding schemes. However, it comes equipped with some pre-cooked recipes that satisfy the immediate needs of the real-world:
+
+### Knowing where tables are
```
this section should probably to go the end of the feature set section? Most people are interested in sharding horizontally first and not vertically? Therefore, I would talk about horizontal sharding features first and then vertical sharding.

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+#### Hashed sharding key
+If the application's sharding key is a monotonically increasing number, then you may not get well-balanced shards. In such cases, you can ask V3 to route queries based on the hash of the main sharding key.
+
+Vitess's filtered replication currently requires that the hash value be physically present as a column in each table. To satisfy this need, you still need to create a column to store this hash value. However, V3 will take care of populating this on your behalf.
```
Maybe put this in parentheses?

You could also slightly reorder the first sentence:

Currently, it is still required to store the hash value in an extra column in addition to the main sharding key because of Vitess' filtered replication feature. V3 will take care of populating the hash column on your behalf.

*The extra hash column will soon be no longer necessary when filtered replication is changed to perform the same hashing.*

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+
+*This restriction will soon be removed once filtered replication is changed to also perform the same hashing.*
+
+#### Auto-increment columns
```
these sections and the following should maybe go into their own section and not in "Sharding schemes" ? More like "Advanced Features" or something even more concrete? I don't know :)

sougou replied:
I think these should be listed here, because this is a features doc. If we were to build a user guide, these topics would belong to the advanced section. Hopefully, if the UI is intuitive enough, we can avoid building the user guide.

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+
+The typical strategy to address this problem is to build a separate lookup table and keep it up-to-date. In the above case, you may build a separate username->user_id relationship table. Once you've informed V3 of this table, it will know what to do with a query like 'select * from user where username=:value'. You can also configure V3 to keep this table up-to-date as you insert or delete data. I ...
+
+*Workflows can be developed to build these indexes on-the-fly.*
```
this isn't very concrete. Be more specific here or omit it?

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+*Workflows can be developed to build these indexes on-the-fly.*
+
+#### Non-unique indexes
+Cross-shard indexes need not be unique. It is possible that rows may exist in multiple shards for a given where clause. V3 allows you to specify indexes as unique or non-unique, and accordingly enforces such constraints during changes.
```
s/need not be/don't have to be/

sounds better?

michael-berlin replied:
not clear from the text what the implications are. Will this feature require an extra mapping table as well?

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+#### Non-unique indexes
+Cross-shard indexes need not be unique. It is possible that rows may exist in multiple shards for a given where clause. V3 allows you to specify indexes as unique or non-unique, and accordingly enforces such constraints during changes.
+
+#### Shared indexes
```
reorder and put this section directly after "Cross-shard indexes" because this section is basically an extension of "Cross-shard indexes".

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+Cross-shard indexes need not be unique. It is possible that rows may exist in multiple shards for a given where clause. V3 allows you to specify indexes as unique or non-unique, and accordingly enforces such constraints during changes.
+
+#### Shared indexes
+There are situations where multiple tables share the same foreign key. A typical use case is a situation where there is a customer table, an order table and an order_detail table. The order table would have a customer_id column. In order to efficiently access all orders of a customer, it would be beneficial to shard this table by customer_id. This will co-locate order rows with their correspondin ...
```
Use three backticks (example: ```TEXT```) to write the table names and column names as code. This way they are easier to distinguish and visually parse from the rest of the words in the sentence.

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+In the case of an order_detail table, it may only need an order_id foreign key. Since this foreign key means the same thing as the order_id in order, creating a cross-shard index for it will result in a duplication of the order_id->customer_id index. In such situations, V3 allows you to just reuse the existing index for the order_detail table also. This saves disk space and also reduces the overa ...
+
+### Consistency
+Once you add multiple indexes to tables, it's possible that the application could make inconsistent requests. V3 makes sure that none of the specified constraints are broken. For example, if a table had both a basic sharding key and a hashed sharding key, it will enforce the rule that the hash of the basic sharding key matches that of the hashed sharding key.
```
> For example, if a table had both a basic sharding key and a hashed sharding key

I don't think this makes sense. You're talking about the case here where we have two columns in the table e.g. "user_id" (basic sharding key) and "keyspace_id" (hashed sharding key)? Earlier you said you want to get rid of that anyway. This would make this feature redundant. Can you maybe give a better example?

michael-berlin replied:
In addition to that, would it make sense to list all use cases? Otherwise this section may be too vague?

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+V3 does not support the full SQL feature set. The current implementation supports simple queries:
+* Single table DML statements: This is a vitess-wide restriction where you can affect only one table and one sharding key per statement. *This restriction may be removed in the future.*
+* Single table SELECT statements:
+  * All constructs allowed if the statement targets only a single sharding key
```
s/allowed/are allowed/

michael-berlin commented on doc/VTGateV3Features.md:
```diff
@@ -0,0 +1,86 @@
+Work is underway to support the following additional constructs:
+* Joins that can be served by sending the query to a single shard.
+* Joins that can be served by sending the query to multiple shards, and trivially combined to form the final result.
+* Cross-shard joins that can be served through a simple nested lookup.
```
What's the difference between this and the previous item? I don't understand it. Maybe reformulate?

michael-berlin commented:
I've read it and left some comments. Feel free to realize them.

sougou commented:
@michael-berlin PTAL.
I believe I've addressed all review comments. I feel like the paragraph hierarchy could be improved. Open to suggestions there.

enisoc commented:
This is going to break some links when it's published. I'll update the links and then run publish.

enisoc commented:
Published.
